### PR TITLE
MDAO find using primary key

### PR DESCRIPTION
**Pull request by amymeng5277**
- [x] 1. Add MDAO testcase 
- [x] 2. Add find_() in AltIndex 
- [x] 3. Refine compare() in PropertyInfo.js

### CODE REVIEW

**Comment by amymeng5277** [comment]
I signed it!

File: src/foam/dao/index/AltIndex.java [hunk 1]
```diff
@@ -55,22 +57,31 @@ public Object remove(Object state, FObject value) {
...
+    }
+    return plan.find(s[0], key);
+  }
+
   public FindPlan planFind(Object state, Object key) {
     Object[] s = toObjectArray(state);
     Plan bestPlan = NoPlan.instance();
 
-    for ( int i = 0 ; i < delegates_.size() ; i++ ) {
+    for (int i = 0; i < delegates_.size(); i++) {
```
Comment (kgrgreer) [inline src/foam/dao/index/AltIndex.java]: Here and on other lines you've broken the FOAM spacing guideline:

https://github.com/foam-framework/foam2/blob/master/doc/guides/StyleGuide.md

File: src/foam/java/PropertyInfo.js [hunk 2]
```diff
@@ -97,7 +97,7 @@ foam.CLASS({
             type: 'int',
             visibility: 'public',
             args: [ { name: 'o1', type: 'Object' }, { name: 'o2', type: 'Object' } ],
-            body: 'return compareValues(get_(o1), get_(o2));'
+            body: 'return (o1 instanceof ' + this.sourceCls.name + ' ? compareValues(get_(o1), get_(o2)) : compareValues(cast(o1), cast(o2)));'
```
Comment (kgrgreer) [inline src/foam/java/PropertyInfo.js]: In JS, we have both compare() and comparePropertyValues().
See: https://github.com/foam-framework/foam2/blob/master/src/foam/core/Property.js#L179

I don't mind also having a compareToPropertyValue or comparePropertyToObj which compares a property value to an object if needed.

File: src/foam/nanos/auth/Country.js [hunk 3]
```diff
@@ -21,5 +21,14 @@ foam.CLASS({
       class: 'String',
       name: 'name'
     }
-  ]
+  ],
+  methods: [{
+    name: 'toString',
+    returns: 'String',
+    javaReturns: 'String',
```
Comment (kgrgreer) [inline src/foam/nanos/auth/Country.js:28]: Does this mess up the JS toString()?

Comment (amymeng5277) [reply src/foam/nanos/auth/Country.js:28]: I don’t think it will mess up js toString, since this is javaCode, right?

**Comment by amymeng5277** [comment]
@kgrgreer  I have fixed all your suggestion please check again.

File: src/foam/java/PropertyInfo.js [hunk 4]
```diff
@@ -103,8 +103,8 @@ foam.CLASS({
             name: 'comparePropertyToObject',
             type: 'int',
             visibility: 'public',
-            args: [ { name: 'key', type: 'Object' }, { name: 'o', type: 'foam.core.FObject' } ],
-            body: 'return compare(cast(key), get_(o));'
+            args: [ { name: 'key', type: 'Object' }, { name: 'o', type: 'Object' } ],
+            body: 'return foam.util.SafetyUtil.compare(cast(key), get_(o));'
```
Comment (kgrgreer) [inline src/foam/java/PropertyInfo.js:107]: I think this could be better. Let's discuss tomorrow.

Comment (kgrgreer) [reply src/foam/java/PropertyInfo.js:107]: I think this line should be: return compareValues(cast(key), get_(o));

**Pull request merged**